kratos/utilities/sparse_matrix_multiplication_utility.h

```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
+                total_nnz += block_nnz;
+            }
+        }
+        KRATOS_ERROR_IF_NOT(check_non_zero == total_nnz) << "Inconsistent number of non-zero values. Check 1: " << total_nnz << " vs " << check_non_zero << std::endl;
+        KRATOS_ERROR_IF_NOT(nonzero_values == total_nnz) << "Inconsistent number of non-zero values. Check 2: " << total_nnz << " vs " << nonzero_values << std::endl;
+    #endif
+
+        // Initialize matrix with the corresponding non-zero values
+        rMatrix = TMatrix(nrows, ncols, nonzero_values);
+
+        // Fill the new matrix
+        double* Matrix_values = rMatrix.value_data().begin();
+        IndexType* Matrix_index1 = rMatrix.index1_data().begin();
+        IndexType* Matrix_index2 = rMatrix.index2_data().begin();
+
+        Matrix_index1[0] = 0;
+        for (IndexType i = 0; i < nrows; ++i)
+            Matrix_index1[i+1] = Matrix_index1[i] + (matrix_ptr[i + 1] - matrix_ptr[i]);
+
+        #pragma omp parallel
+        {
+            #pragma omp for
+            for (int i=0; i<static_cast<int>(number_of_rows_blocks); ++i) {
+                for (int k=0; k<static_cast<int>(row_sizes[i]); ++k) {
+                    const IndexType row_beg = matrix_ptr[std::accumulate(row_sizes.begin(), row_sizes.begin() + i, 0) + k];
+                    IndexType row_end = row_beg;
+                    for (int j=0; j<static_cast<int>(number_of_columns_blocks); ++j) {
+                        const SizeType initial_index_column = std::accumulate(column_sizes.begin(), column_sizes.begin() + j, 0);
+
+                        // Skip if empty matrix
+                        TMatrix& r_matrix = *rMatricespBlocks(i, j);
+                        if (r_matrix.nnz() > 0) {
+                            if (TransposeBlocks(i, j)) {
+                                // We compute the transposed matrix
+                                const SizeType size_system_1 = r_matrix.size1();
+                                const SizeType size_system_2 = r_matrix.size2();
+                                TMatrix transpose(size_system_2, size_system_1);
+                                TransposeMatrix<TMatrix, TMatrix>(transpose, r_matrix);
+                                ComputeAuxiliarValuesBlocks<TMatrix>(transpose, Matrix_index2, Matrix_values, k, row_end, initial_index_column, ContributionCoefficients(i, j));
+                            } else {
+                                ComputeAuxiliarValuesBlocks<TMatrix>(r_matrix, Matrix_index2, Matrix_values, k, row_end, initial_index_column, ContributionCoefficients(i, j));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Close the matrix
+        rMatrix.set_filled(nrows+1, nonzero_values);
```

Comment:
also this is specific of compressed_matrix

Reply:
See, the rest of the sparse_matrix_multiplication_utility.h  is designed for compressed_matrix. This is consistent with what it is in the rest of the utility. Additionally the B&S consider also to work with compressed_matrix and we still consider the space as a template argument. Should we refactor all these implementations first? 

For example: https://github.com/KratosMultiphysics/Kratos/blob/0c0a3dbac08f886faca926d11c0b049206d7cfd8/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1253 and https://github.com/KratosMultiphysics/Kratos/blob/0c0a3dbac08f886faca926d11c0b049206d7cfd8/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L1280